File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKActionMoveTo.java
package br.com.insanitech.spritekit;

/**
 * Created by anderson on 06/01/17.
 */

class SKActionMoveTo extends SKAction {
    private SKPoint startPoint = new SKPoint();
    private SKPoint deltaPoint = new SKPoint();

    public SKActionMoveTo(SKPoint delta) {
        this.deltaPoint.assignByValue(delta);
    }

    @Override
    void computeStart() {
        startPoint.assignByValue(getParent().getPosition());
    }

    @Override
    void computeAction(long elapsed) {
        float newX = 0.0f;
        float newY = 0.0f;
        switch (timingMode) {
            case EaseIn: {
                newX = SKEaseCalculations.easeIn(elapsed, startPoint.x, deltaPoint.x - startPoint.x, duration);
                newY = SKEaseCalculations.easeIn(elapsed, startPoint.y, deltaPoint.y - startPoint.y, duration);
            }
            break;

            case EaseInEaseOut: {
                newX = SKEaseCalculations.easeInOut(elapsed, startPoint.x, deltaPoint.x - startPoint.x, duration);
                newY = SKEaseCalculations.easeInOut(elapsed, startPoint.y, deltaPoint.y - startPoint.y, duration);
            }
            break;

            case EaseOut: {
                newX = SKEaseCalculations.easeOut(elapsed, startPoint.x, deltaPoint.x - startPoint.x, duration);
                newY = SKEaseCalculations.easeOut(elapsed, startPoint.y, deltaPoint.y - startPoint.y, duration);
            }
            break;

            case Linear: {
                newX = SKEaseCalculations.linear(elapsed, startPoint.x, deltaPoint.x - startPoint.x, duration);
                newY = SKEaseCalculations.linear(elapsed, startPoint.y, deltaPoint.y - startPoint.y, duration);
            }
            break;
        }
        getParent().setPosition(newX, newY);
    }

    @Override
    void computeFinish() {
        getParent().setPosition(deltaPoint.x, deltaPoint.y);
    }

    @Override
    boolean willHandleFinish() {
        return false;
    }
}

File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKActionMoveTo.kt
package br.com.insanitech.spritekit

/**
 * Created by anderson on 06/01/17.
 */

internal class SKActionMoveTo(delta: SKPoint) : SKAction() {
    private val startPoint = SKPoint()
    private val deltaPoint = SKPoint()

    init {
        this.deltaPoint.assignByValue(delta)
    }

    internal override fun computeStart() {
        (parent ?: return).let { parent ->
            startPoint.assignByValue(parent.position)
        }
    }

    internal override fun computeAction(elapsed: Long) {
        var newX = 0.0f
        var newY = 0.0f
        when (timingMode) {
            SKActionTimingMode.EaseIn -> {
                newX = SKEaseCalculations.easeIn(elapsed.toFloat(), startPoint.x, deltaPoint.x - startPoint.x, duration.toFloat())
                newY = SKEaseCalculations.easeIn(elapsed.toFloat(), startPoint.y, deltaPoint.y - startPoint.y, duration.toFloat())
            }

            SKActionTimingMode.EaseInEaseOut -> {
                newX = SKEaseCalculations.easeInOut(elapsed.toFloat(), startPoint.x, deltaPoint.x - startPoint.x, duration.toFloat())
                newY = SKEaseCalculations.easeInOut(elapsed.toFloat(), startPoint.y, deltaPoint.y - startPoint.y, duration.toFloat())
            }

            SKActionTimingMode.EaseOut -> {
                newX = SKEaseCalculations.easeOut(elapsed.toFloat(), startPoint.x, deltaPoint.x - startPoint.x, duration.toFloat())
                newY = SKEaseCalculations.easeOut(elapsed.toFloat(), startPoint.y, deltaPoint.y - startPoint.y, duration.toFloat())
            }

            SKActionTimingMode.Linear -> {
                newX = SKEaseCalculations.linear(elapsed.toFloat(), startPoint.x, deltaPoint.x - startPoint.x, duration.toFloat())
                newY = SKEaseCalculations.linear(elapsed.toFloat(), startPoint.y, deltaPoint.y - startPoint.y, duration.toFloat())
            }
        }
        parent?.setPosition(newX, newY)
    }

    internal override fun computeFinish() {
        parent?.setPosition(deltaPoint.x, deltaPoint.y)
    }
}
